File: typescript/src/hip3Order.ts
import { exchange, signHash, getHip3Mid } from "./client";

const COIN = "xyz:SILVER";

async function main() {
  let mid = await getHip3Mid(COIN);
  if (mid === 0) {
    console.log(`Could not fetch ${COIN} mid price, using fallback`);
    mid = 78.0;
  }

  const sz = (11.0 / mid).toFixed(2);
  const buyPx = (mid * 1.03).toFixed(2);

  console.log(`${COIN} mid: $${mid.toFixed(2)}`);
  console.log(`BUY ${sz} @ ${buyPx} (IOC, ~$${(parseFloat(sz) * mid).toFixed(2)} notional)\n`);

  const res = await exchange({
    action: {
      type: "order",
      orders: [{ asset: COIN, side: "buy", price: buyPx, size: sz, tif: "ioc" }],
    },
  });
  const sig = await signHash(res.hash);

  const result = await exchange({
    action: res.action,
    nonce: res.nonce,
    signature: sig,
  });

  console.log(JSON.stringify(result.exchangeResponse, null, 2));
}

main();

File: python/hip3_order.py
"""Place an IOC order on a HIP-3 market by name (e.g. xyz:SILVER)."""

import json
from client import exchange, sign_hash, get_hip3_mid

COIN = "xyz:SILVER"

mid = get_hip3_mid(COIN)
if mid == 0:
    print(f"Could not fetch {COIN} mid price, using fallback")
    mid = 78.0

sz = round(10.0 / mid, 2)
buy_px = round(mid * 1.03, 2)

print(f"{COIN} mid: ${mid:,.2f}")
print(f"BUY {sz} @ {buy_px} (IOC, ~${sz * mid:.2f} notional)\n")

res = exchange({
    "action": {
        "type": "order",
        "orders": [{
            "asset": COIN,
            "side": "buy",
            "price": str(buy_px),
            "size": str(sz),
            "tif": "ioc",
        }],
    },
})
sig = sign_hash(res["hash"])

result = exchange({
    "action": res["action"],
    "nonce": res["nonce"],
    "signature": sig,
})

print(json.dumps(result["exchangeResponse"], indent=2))
